src/gui/sdlmain.cpp

```diff
@@ -773,17 +786,104 @@ static bool LoadGLShaders(const char *src, GLuint *vertex, GLuint *fragment) {
 }
 #endif
 
-Bitu GFX_SetSize(Bitu width,Bitu height,Bitu flags,double scalex,double scaley,GFX_CallBack_t callback) {
+/* Get the maximum area available for drawing: */
+/* TODO: The current implementation will not let the pixel-perfect mode to  */
+/*       to apply even the simplest 1:2 and 2:1 corrections at the original */
+/*       window resolution, therefore: consider a special case for PP mode. */
+static void GetAvailableArea(Bit16u *width, Bit16u *height)
+{
+	bool   fixed;
+	double par;
+
+	fixed = false;
+	if (sdl.desktop.fullscreen) {
+		if (sdl.desktop.full.fixed) {
+			*width  = sdl.desktop.full.width;
+			*height = sdl.desktop.full.height;
+			fixed   = true;
+		}
+	}
+	else {
+		if (sdl.desktop.window.width > 0) {
+			*width  = sdl.desktop.window.width;
+			*height = sdl.desktop.window.height;
+			fixed   = true;
+		}
+	}
+	if (!fixed) {
+		par = sdl.draw.par;
+		if (sdl.draw.par > 1.0) *height = round( ( double )*height * par );
+		if (sdl.draw.par < 1.0) *width  = round( ( double )*width  / par ); 
+	}
+}
```

Comment:
Oh, I would also join line 805 and 806: `} else {`.

I have some concerns about the code in those ifs 799-812, but I need to read into the code to tell for sure (I'm concerned about `fullresolution=desktop`).

Reply:
*dreamer*:

> Very nicely formatted function, thank you! :)

Thanks, but I don't like it at all :-)

> Change input param types to `uint16_t`, please. In general we need to move away from using 16-bit width/height and move towards `int`, as used in SDL2 - but it will be a long process; you can keep using 16-bit for now, but let's avoid `Bit16u`.

May I change it to `int`?

> Line 795: `bool fixed = false`;
> 
> Move declaration of variable `par` to line 814 (after moving it, short name will be ok).

I think that mixing declaration and assignment in one statement is wrong because they are two completely different operations that relate as metadata and data. Doing that all over the code instead of at the top of function thwarts readability because executable code (actions, assignments) is interrupted but declarations. It reads smoother with the declarations at the top.

*Bart* has come up with [more](https://pastebin.com/raw/3CwyhFEz) reasons.

Furthermore, this approach facilitates refactoring: once the need to scroll up in order to add another variable starts to be annoying, it is a veritable indication that the function should be split into several smaller ones.
 
> Lines 815 and 816 have some issues, they introduce useless casting and are lacking proper casting of result;

Thanks. I have removed the useless `cast`s. But why does one need to cast the result explicitly? Is there any situation when it will affect the result?

> I suggest rewriting them this way:
> 
> ```
> if (sdl.draw.pixel_aspect_ratio > 1.0)
> 	*height = static_cast<uint16_t>(round((*height) * par));
> ```

I have rewritten my two lines as four lines.

Reply:
By the way, consider my compact code:

```
   if (!fixed) {
      par = sdl.draw.par;
      if (par > 1.0) *height = round( *height * par );
      if (par < 1.0) *width  = round( *width  / par );
   }
```
I store the value of `sdl.draw.par` into `par` to make the code shorter. If we introduce long names such as `pixel_aspect_ratio` such will become impossible or will require even more dummy assignment to give a shorter name to a local variable within a small scope...

Reply:
> Oh, I would also join line 805 and 806: `} else {`.

It would make that `else` align with wrong `if`. I should prefer to keep it on a separate line if possible.

> I have some concerns about the code in those ifs 799-812, but I need to read into the code to tell for sure (I'm concerned about `fullresolution=desktop`).

That code adjusts `width` and `height` in fullscreen mode if fullscreen resolution differs from the original resolution of the game.